test/lint/lint-files.py

```diff
@@ -0,0 +1,207 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+"""
+This checks that all files in the repository have correct filenames and permissions
+"""
+
+import os
+import re
+import sys
+from subprocess import check_output
+from typing import Optional, NoReturn
+
+CMD_ALL_FILES = "git ls-files --full-name"
+CMD_SOURCE_FILES = 'git ls-files --full-name -- "*.[cC][pP][pP]" "*.[hH]" "*.[pP][yY]" "*.[sS][hH]"'
+CMD_SHEBANG_FILES = "git grep --full-name --line-number -I '^#!'"
+ALLOWED_FILENAME_REGEXP = "^[a-zA-Z0-9/_.@][a-zA-Z0-9/_.@-]*$"
+ALLOWED_SOURCE_FILENAME_REGEXP = "^[a-z0-9_./-]+$"
+ALLOWED_SOURCE_FILENAME_EXCEPTION_REGEXP = (
+    "^src/(secp256k1/|univalue/|test/fuzz/FuzzedDataProvider.h)"
+)
+ALLOWED_PERMISSION_NON_EXECUTABLES = 644
+ALLOWED_PERMISSION_EXECUTABLES = 755
+ALLOWED_EXECUTABLE_SHEBANG = {
+    "py": [b"#!/usr/bin/env python3"],
+    "sh": [b"#!/usr/bin/env bash", b"#!/bin/sh"],
+}
+
+
+class FileMeta(object):
+    def __init__(self, file_path: str):
+        self.file_path = file_path
+
+    @property
+    def extension(self) -> Optional[str]:
+        """
+        Returns the file extension for a given filename string.
+        eg:
+        'ci/lint_run_all.sh' -> 'sh'
+        'ci/retry/retry' -> None
+        'contrib/devtools/split-debug.sh.in' -> 'in'
+        """
+        return str(os.path.splitext(self.file_path)[1].strip(".") or None)
+
+    @property
+    def full_extension(self) -> Optional[str]:
+        """
+        Returns the full file extension for a given filename string.
+        eg:
+        'ci/lint_run_all.sh' -> 'sh'
+        'ci/retry/retry' -> None
+        'contrib/devtools/split-debug.sh.in' -> 'sh.in'
+        """
+        filename_parts = self.file_path.split(os.extsep, 1)
+        try:
+            return filename_parts[1]
+        except IndexError:
+            return None
+
+    @property
+    def permissions(self) -> int:
+        """
+        Returns the octal file permission of the file
+        """
+        return int(oct(os.stat(self.file_path).st_mode)[-3:])
+
+
+def check_all_filenames() -> int:
+    """
+    Checks every file in the repository against an allowed regexp to make sure only lowercase or uppercase
+    alphanumerics (a-zA-Z0-9), underscores (_), hyphens (-), at (@) and dots (.) are used in repository filenames.
+    """
+    # We avoid using rstrip() to ensure we catch filenames which accidentally include trailing whitespace
+    filenames = check_output(CMD_ALL_FILES, shell=True).decode("utf8").split("\n")
+    filenames = [filename for filename in filenames if filename != ""]  # removes the trailing empty list element
```

Comment:
nit: The comment does not correspond with the code actually. 

One could probably do: `filenames = filenames[:-1]`

Reply:
right, I chose to remove the trailing element by checking for an empty string as I was worried about some case in the future where the `git ls-files` command changes somehow and the last element is not actually empty. I'm probably overthinking it, though.

Reply:
On a review of this, I'm fairly certain I can just safely rewrite this code as this one liner, which still catches the original edge case of a file with whitespace at the end of the files list.

```python
filenames = check_output(CMD_ALL_FILES, shell=True).decode("utf8").rstrip('\n').split("\n")
```

Happy to update this if people don't mind re-reviewing. Let me know.

Reply:
> right, I chose to remove the trailing element by checking for an empty string as I was worried about some case in the future where the output of the `git ls-files` command changes somehow and the last element is not actually empty. I'm probably overthinking it though.

One would need to ask on a Git mailing list whether the text output of the command is stable or not. But your solution works well, so it's good as it is I think. 👍

Reply:
If a filename had a newline in its name it would also be split by the `split("\n")`

Reply:
I'll think about a way to handle this case. I think currently if this happens the test overall will still fail, as the permissions test would not be able to locate the file and should error out - we can definitely handle this better, though.

Reply:
I think `.rstrip('\n').split("\n")` as you suggested is fine

Reply:
Addressed in https://github.com/bitcoin/bitcoin/pull/21873